plugins/redis/redis.cpp

```diff
@@ -375,10 +433,245 @@ void Connection::parseOptions(ICodeContext * ctx, const char * _options)
+    return (reply->type == REDIS_REPLY_ARRAY)
+            && (reply->elements > 1)
+            && (strcmp(op, reply->element[0]->str) == 0)
+            && (strcmp(channel.str(), reply->element[1]->str) == 0);
+}
+void Connection::subscribe(ICodeContext * ctx, const char * _channel)
+{
+    assert(!subscribed || channel.isEmpty());//If this is a reused subscriptionConnection then this implies that it was not correctly unsubscribed/reset.
+    channel.set(_channel);
+
+    OwnedReply reply = Reply::createReply(redisCommand("SUBSCRIBE %b", channel.str(), (size_t)channel.length()));
+    /* It is possible for the subscription to succeed in being registered by the server but fails in receiving the associated reply.
+     * Set early to prevent not unsubscribing when failing.
+     */
+    subscribed = true;
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+    if (!isCorrectChannel(reply->query(), "subscribe"))
+        fail("SUBSCRIBE", "failed to register SUB", channel.str());
+}
+void Connection::subAndWaitForSinglePub(ICodeContext * ctx, const char * keyOrChannel, size_t & messageSize, char * & message, int _database, bool lockedKey)
+{
+    StringBuffer channelBuffer;
+    encodeChannel(channelBuffer, keyOrChannel, _database, lockedKey);
+    subscribe(ctx, channelBuffer.str());
+
+    //Now wait for published message
+    OwnedReply reply = new Reply();
+    readReply(reply);
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+
+    if (isCorrectChannel(reply->query(), "message"))
+    {
+        if (reply->query()->element[2]->len > 0)
+        {
+            messageSize = (size_t)reply->query()->element[2]->len;
+            message = reinterpret_cast<char*>(allocateAndCopy(reply->query()->element[2]->str, messageSize));
+        }
+        else
+        {
+            messageSize = 0;
+            message = nullptr;
+        }
+        unsubscribe(ctx, nullptr);
+        return;
+    }
+    throwUnexpected();
+}
+int Connection::writeBufferToSocket()
+{
+    int done = 0;
+    void * aux = nullptr;
+    if (context->flags & REDIS_BLOCK)
+    {
+       /* Write until done */
+       while (!done)
+       {
+           if (redisBufferWrite(context, &done) == REDIS_ERR)
+               return REDIS_ERR;
+       }
+    }
+    return REDIS_OK;
+}
+unsigned Connection::getUnsubscribeReadAttempts()
+{
+    if (!unsubscribeReadAttemptsChecked)//unsubscribeReadAttemptsChecked is std:atomic<bool>. Test to guard against unnecessary critical section
+    {
+        CriticalBlock block(critsec);
+        if (!unsubscribeReadAttemptsChecked)
+        {
+            const char * tmp = getenv("HPCC_REDIS_PLUGIN_UNSUBSCRIBE_READ_ATTEMPTS");//ms
+            if (tmp && *tmp)
+                unsubscribeReadAttempts = atoi(tmp);
+            unsubscribeReadAttemptsChecked = true;
+        }
+    }
+    return unsubscribeReadAttempts;
+}
+void Connection::unsubscribe(ICodeContext * ctx, const char * password)
```

Comment:
In a previous instance, the pattern was not as simple as to just free the context if it failed to unsubscribe and would, instead, as an option at least, re-connect there and then. To do so the password is required. I had left in to make any future changes easier in case these are needed and thus not having to significantly refactor the code to account for the parameter propagation. I will remove though.